### ISSUE TITLE
greyscale() sets wrong OutputInfo.channels

### ISSUE THREAD

**Issue by koffmoff** [lovell/sharp#1425]
After doing a grayscale() operation, OutputInfo.channels = 3, when it really should be 1.

Example:

```ts
const logo = await logoSrc
    .greyscale()
    .raw()
    .toBuffer({ resolveWithObject: true })
// This is set to 3, which is wrong. It should be 1 since the picture is grayscale. Have to manually adjust it to 1.
logo.info.channels = 1
```

**Comment by lovell** [comment]
Hello, did you see the [greyscale](http://sharp.pixelplumbing.com/en/stable/api-colour/#greyscale) docs?

> "By default the output image will be web-friendly sRGB and contain three (identical) color channels. This may be overridden by other sharp operations such as toColourspace('b-w'), which will produce an output image containing one color channel."

**Comment by koffmoff** [comment]
Yes, but I'm not sure it's correct. Look at this:

```ts
const gray = await sharp({create: {width: 100, height: 100, channels: 3, background: {r: 255, g: 255, b: 255}}})
    .greyscale()
    .raw()
    .toBuffer({resolveWithObject: true})
// Comment out this line to get: VipsImage: memory area too small --- should be 30000 bytes, you passed 10000
gray.info.channels = 1
const out = await sharp(gray.data, {raw: {width: gray.info.width, height: gray.info.height, channels: gray.info.channels}})
    .raw()
    .toBuffer()
````

After the first operation, gray.info.channels = 3. Ok. As per the docs. Problem is, gray.data.length = 10000. And that's just one channel (100x100), or what am I getting wrong? Forcing it to one channel makes the second operaton work.

**Comment by lovell** [comment]
Yes, sorry, using `raw` output extracts a single channel so this is a bug.

Happy to accept and/or help with a PR if you're able to add the fix and a test case. `baton->channels` needs to be modified inside the `if` block approximately here:

https://github.com/lovell/sharp/blob/de11d36d004b69e16d58e77d0b92c7897386b55d/src/pipeline.cc#L771-L774

**Comment by lovell** [comment]
Commit 6f9699f adds a test assertion that would have caught this, plus fixes the bug. This will be in v0.21.1, thanks for reporting.
